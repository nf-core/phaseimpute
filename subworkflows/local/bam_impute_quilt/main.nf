include { QUILT_QUILT                        } from '../../../modules/nf-core/quilt/quilt'
include { BCFTOOLS_INDEX                     } from '../../../modules/nf-core/bcftools/index'
include { BCFTOOLS_ANNOTATE                  } from '../../../modules/nf-core/bcftools/annotate'

workflow BAM_IMPUTE_QUILT {

    take:
    ch_input             // channel: [ [id], bam, bai ]
    ch_hap_legend        // channel: [ [panel, chr], hap, legend ]
    ch_chunks            // channel: [ [panel, chr], start_coordinate, end_coordinate, number ]


    main:


    ch_versions = Channel.empty()

    posfile             = []
    phasefile           = []
    posfile_phasefile   = [[id: null], posfile, phasefile]
    genetic_map_file    = []
    fasta               = [[id:'test'], []]

    ngen                = params.ngen
    buffer              = params.buffer

    if (genetic_map_file.isEmpty()) {
        ch_hap_chunks = ch_hap_legend.combine(ch_chunks, by:0).map { it + ngen + buffer + [[]] }
    } else {
        // Add ngen and buffer + genetic map file (untested)
        ch_hap_chunks = ch_hap_legend.combine(ch_chunks, by:0).join(genetic_map_file)
    }

    ch_quilt = ch_input
        .combine(ch_hap_chunks)
        .map {
            metaIC, bam, bai, metaPC, hap, legend, chr, start, end, ngen, buffer, gmap ->
            [
                metaIC.subMap("id") + ["panel": metaPC.id, "chr": metaPC.chr, "chunk": start + "-" + end],
                bam, bai, hap, legend, chr, start, end, ngen, buffer, gmap
            ]
        }

    // Run QUILT
    QUILT_QUILT ( ch_quilt, posfile_phasefile, fasta )
    ch_versions = ch_versions.mix(QUILT_QUILT.out.versions.first())

    // Index imputed VCF
    BCFTOOLS_INDEX(QUILT_QUILT.out.vcf)
    ch_versions = ch_versions.mix(BCFTOOLS_INDEX.out.versions.first())

    // Annotate the variants
    BCFTOOLS_ANNOTATE(QUILT_QUILT.out.vcf
        .join(BCFTOOLS_INDEX.out.tbi)
        .combine(Channel.of([[], [], [], []]))
    )
    ch_versions = ch_versions.mix(BCFTOOLS_ANNOTATE.out.versions.first())

    // Join VCFs and TBIs
    ch_vcf_tbi = BCFTOOLS_ANNOTATE.out.vcf
        .join(BCFTOOLS_ANNOTATE.out.tbi)
        .map { metaIPC, vcf, tbi -> [metaIPC + [tools: "quilt"], vcf, tbi] }

    emit:
    vcf_tbi     = ch_vcf_tbi               // channel:  [ [id, panel], vcf, tbi ]
    versions    = ch_versions              // channel:  [ versions.yml ]
}
