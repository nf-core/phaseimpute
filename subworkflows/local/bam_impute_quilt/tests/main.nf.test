nextflow_workflow {

    name "Test Subworkflow BAM_IMPUTE_QUILT"
    script "../main.nf"
    config "./nextflow.config"

    workflow "BAM_IMPUTE_QUILT"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/bam_impute_quilt"
    tag "bam_impute_quilt"

    tag "quilt"
    tag "quilt/quilt"
    tag "bcftools"
    tag "bcftools/index"
    tag "bcftools/annotate"

    test("Impute with quilt") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
            }
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "NA12878"],
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam.bai", checkIfExist:true),
                    ],
                    [
                        [id: "NA19401"],
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam.bai", checkIfExist:true),
                    ],
                ])
                input[1] = Channel.of([
                        [id: "1000GP", chr: "22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.hap.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.legend.gz", checkIfExist:true)
                    ]).collect()
                input[2] = Channel.of(
                    [[chr: "22", id: "1000GP"], "chr22", "16570065", "16597215"],
                    [[chr: "22", id: "1000GP"], "chr22", "16587172", "16609999"]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                { assert snapshot(workflow.out.vcf_tbi.collect{ path(it[1]).vcf.summary }).match("Summary") }
            )
        }
    }

    test("Impute with quilt mixed") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
            }
            workflow {
                """
                input[0] = Channel.of([
                    [id: "allIndividuals"],
                    [file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam", checkIfExist:true),
                    file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam", checkIfExist:true)],
                    [file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam.bai", checkIfExist:true),
                    file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam.bai", checkIfExist:true)],
                ])
                input[1] = Channel.of([
                        [id: "1000GP", chr: "22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.hap.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.legend.gz", checkIfExist:true)
                    ]).collect()
                input[2] = Channel.of(
                    [[chr: "22", id: "1000GP"], "chr22", "16570065", "16597215"],
                    [[chr: "22", id: "1000GP"], "chr22", "16587172", "16609999"]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                { assert snapshot(workflow.out.vcf_tbi.collect{ path(it[1]).vcf.summary }).match("SummaryMixed") }
            )
        }
    }
}
