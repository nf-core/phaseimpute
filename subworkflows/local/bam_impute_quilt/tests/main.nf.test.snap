{
    "Impute with quilt mixed": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "allIndividuals",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16570065-16597215",
                            "tools": "quilt"
                        },
                        "allIndividuals_22_16570065-16597215.impute.annotate.vcf.gz:md5,37547c230e2ca2c4685b748e9cf33ee0",
                        "allIndividuals_22_16570065-16597215.impute.annotate.vcf.gz.tbi:md5,7f282ce0da4e46c8e22258fe4f04535f"
                    ],
                    [
                        {
                            "id": "allIndividuals",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16587172-16609999",
                            "tools": "quilt"
                        },
                        "allIndividuals_22_16587172-16609999.impute.annotate.vcf.gz:md5,39adb2163103938396050e12808f073c",
                        "allIndividuals_22_16587172-16609999.impute.annotate.vcf.gz.tbi:md5,3b1912e53aa488aaa05a6f56c17a3ed5"
                    ]
                ],
                "1": [
                    "versions.yml:md5,ac2d0f2a8e77cd796d3b4b84fdd46b0d",
                    "versions.yml:md5,cb038a1f67a47a76d0e0a983cfd56789"
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "allIndividuals",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16570065-16597215",
                            "tools": "quilt"
                        },
                        "allIndividuals_22_16570065-16597215.impute.annotate.vcf.gz:md5,37547c230e2ca2c4685b748e9cf33ee0",
                        "allIndividuals_22_16570065-16597215.impute.annotate.vcf.gz.tbi:md5,7f282ce0da4e46c8e22258fe4f04535f"
                    ],
                    [
                        {
                            "id": "allIndividuals",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16587172-16609999",
                            "tools": "quilt"
                        },
                        "allIndividuals_22_16587172-16609999.impute.annotate.vcf.gz:md5,39adb2163103938396050e12808f073c",
                        "allIndividuals_22_16587172-16609999.impute.annotate.vcf.gz.tbi:md5,3b1912e53aa488aaa05a6f56c17a3ed5"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,ac2d0f2a8e77cd796d3b4b84fdd46b0d",
                    "versions.yml:md5,cb038a1f67a47a76d0e0a983cfd56789"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T16:12:41.131269971"
    },
    "Summary": {
        "content": [
            [
                "VcfFile [chromosomes=[22], sampleCount=1, variantCount=572, phased=true, phasedAutodetect=false]",
                "VcfFile [chromosomes=[22], sampleCount=1, variantCount=570, phased=true, phasedAutodetect=false]",
                "VcfFile [chromosomes=[22], sampleCount=1, variantCount=572, phased=true, phasedAutodetect=false]",
                "VcfFile [chromosomes=[22], sampleCount=1, variantCount=570, phased=true, phasedAutodetect=false]"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T16:06:47.574062814"
    },
    "Impute with quilt": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16570065-16597215",
                            "tools": "quilt"
                        },
                        "NA12878_22_16570065-16597215.impute.annotate.vcf.gz:md5,857e034286763a2316a68f7314786dda",
                        "NA12878_22_16570065-16597215.impute.annotate.vcf.gz.tbi:md5,de3b20dc5f67aae2a9669f0c667edb5d"
                    ],
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16587172-16609999",
                            "tools": "quilt"
                        },
                        "NA12878_22_16587172-16609999.impute.annotate.vcf.gz:md5,2c3995176d978021a797def9951a1529",
                        "NA12878_22_16587172-16609999.impute.annotate.vcf.gz.tbi:md5,674782f71cca76015e0b72f9bc4e5431"
                    ],
                    [
                        {
                            "id": "NA19401",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16570065-16597215",
                            "tools": "quilt"
                        },
                        "NA19401_22_16570065-16597215.impute.annotate.vcf.gz:md5,f3efd22b0b87e1cd98ae75c727b787dc",
                        "NA19401_22_16570065-16597215.impute.annotate.vcf.gz.tbi:md5,b46f4fb03214dd4b09e4144dc8b994b7"
                    ],
                    [
                        {
                            "id": "NA19401",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16587172-16609999",
                            "tools": "quilt"
                        },
                        "NA19401_22_16587172-16609999.impute.annotate.vcf.gz:md5,b7258fcd25e0140da3ba48c07410d41a",
                        "NA19401_22_16587172-16609999.impute.annotate.vcf.gz.tbi:md5,889c40d51c9531df56404db6d3025c70"
                    ]
                ],
                "1": [
                    "versions.yml:md5,ac2d0f2a8e77cd796d3b4b84fdd46b0d",
                    "versions.yml:md5,cb038a1f67a47a76d0e0a983cfd56789"
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16570065-16597215",
                            "tools": "quilt"
                        },
                        "NA12878_22_16570065-16597215.impute.annotate.vcf.gz:md5,857e034286763a2316a68f7314786dda",
                        "NA12878_22_16570065-16597215.impute.annotate.vcf.gz.tbi:md5,de3b20dc5f67aae2a9669f0c667edb5d"
                    ],
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16587172-16609999",
                            "tools": "quilt"
                        },
                        "NA12878_22_16587172-16609999.impute.annotate.vcf.gz:md5,2c3995176d978021a797def9951a1529",
                        "NA12878_22_16587172-16609999.impute.annotate.vcf.gz.tbi:md5,674782f71cca76015e0b72f9bc4e5431"
                    ],
                    [
                        {
                            "id": "NA19401",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16570065-16597215",
                            "tools": "quilt"
                        },
                        "NA19401_22_16570065-16597215.impute.annotate.vcf.gz:md5,f3efd22b0b87e1cd98ae75c727b787dc",
                        "NA19401_22_16570065-16597215.impute.annotate.vcf.gz.tbi:md5,b46f4fb03214dd4b09e4144dc8b994b7"
                    ],
                    [
                        {
                            "id": "NA19401",
                            "panel": "1000GP",
                            "chr": "22",
                            "chunk": "16587172-16609999",
                            "tools": "quilt"
                        },
                        "NA19401_22_16587172-16609999.impute.annotate.vcf.gz:md5,b7258fcd25e0140da3ba48c07410d41a",
                        "NA19401_22_16587172-16609999.impute.annotate.vcf.gz.tbi:md5,889c40d51c9531df56404db6d3025c70"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,ac2d0f2a8e77cd796d3b4b84fdd46b0d",
                    "versions.yml:md5,cb038a1f67a47a76d0e0a983cfd56789"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T16:12:02.559345299"
    },
    "SummaryMixed": {
        "content": [
            [
                "VcfFile [chromosomes=[22], sampleCount=2, variantCount=572, phased=true, phasedAutodetect=false]",
                "VcfFile [chromosomes=[22], sampleCount=2, variantCount=570, phased=true, phasedAutodetect=false]"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T16:12:41.764403087"
    }
}